runtime/lua/vim/ui/img/utils/screen.lua

```diff
@@ -0,0 +1,174 @@
+  if vim.fn.has('win32') == 1 then
+    size = M.__csi_size()
+  elseif vim.fn.has('unix') == 1 then
+    size = M.__posix_size()
+  end
+
+  M.__size = size or M.__default_size()
+
+  return M.__size
+end
+
+---@private
+---Determines the size of the terminal screen using CSI escape codes.
+---@return vim.ui.img.utils.ScreenSize|nil
+function M.__csi_size()
+  -- TODO: Introduce support for querying CSI. Neovim eats the response right now.
+  --
+  --       CSI 14 t and CSI 16 t are both supported by Windows Terminal
+  --
+  --       CSI 14 t :: tells us the pixel dimensions of the view space
+  --       CSI 16 t :: tells us the pixel dimensions of a terminal character
+  vim.notify(
+    'support querying CSI 14 t and/or CSI 16 t is not available',
+    vim.log.levels.WARN
+  )
+  return nil
+end
+
+---@private
+---Determines the size of the terminal screen for POSIX systems.
+---@return vim.ui.img.utils.ScreenSize|nil
+function M.__posix_size()
+  ---@type vim.ui.img.utils.ScreenSize|nil
+  local size
+
+  -- On Linux/Android, BSD, MacOS, and Solaris we use
+  -- ioctl with TIOCGWINSZ to calculate the size.
+  --
+  -- Because of this, we define a structure to collect
+  -- the size information and specify ioctl as available.
+  local ffi = require('ffi')
+  if not M.__def then
+    ffi.cdef([[
+      typedef struct {
+        unsigned short row;
+        unsigned short col;
+        unsigned short xpixel;
+        unsigned short ypixel;
+      } winsize;
+      int ioctl(int, int, ...);
+    ]])
+    M.__def = true
+  end
+
+  local TIOCGWINSZ = nil
+
+  -- 1. For Linux, Android, GNU Hurd, and WSL
+  -- 2. For MacOS and BSD like FreeBSD, OpenBSD, NetBSD, and DragonflyBSD
+  -- 3. For Solaris
+  if vim.fn.has('linux') == 1 then
+    TIOCGWINSZ = 0x5413
+  elseif vim.fn.has('mac') == 1 or vim.fn.has('bsd') == 1 then
+    TIOCGWINSZ = 0x40087468
+  elseif vim.fn.has('sun') == 1 then
+    TIOCGWINSZ = 0x5468
+  end
+
+  ---@type boolean, string|nil
+  local ok, err = pcall(function()
+    ---@type { row: number, col: number, xpixel: number, ypixel: number }
+    local sz = ffi.new('winsize')
+    if ffi.C.ioctl(1, TIOCGWINSZ, sz) ~= 0 or sz.col == 0 or sz.row == 0 then
+      return
+    end
+    size = {
+      width = sz.xpixel,
+      height = sz.ypixel,
+      columns = sz.col,
+      rows = sz.row,
+      cell_width = sz.xpixel / sz.col,
+      cell_height = sz.ypixel / sz.row,
+      scale = math.max(1, sz.xpixel / sz.col / 8),
+    }
+  end)
+
+  if not ok then
+    vim.notify(
+      string.format('unable to retrieve screen size (POSIX): %s', err or ''),
+      vim.log.levels.WARN
+    )
+  end
+
+  return size
+end
+
+---@private
+---@return vim.ui.img.utils.ScreenSize
+function M.__default_size()
+  local cell_width = 9
+  local cell_height = 18
+
+  -- Assume that for Windows the 10x20 used by Windows Terminal
+  -- is a better choice for a default
+  if vim.fn.has('win32') == 1 then
+    cell_width = 10
+    cell_height = 20
+  end
```

Comment:
I'm guessing this is just testing whether vim itself is running on Windows, which isn't particularly meaningful. If you're using a Windows terminal you're just as likely (if not more likely) to be running vim in WSL. or when connected to a Linux server. And you'd assumedly also get a false positive with a Linux terminal remoting onto a Windows server.

I thought this would be an easy win, but if it's a problem to make 10x20 the universal default, it's probably best to just drop it for now, and wait for queries to be fixed.

Reply:
I can set the whole thing to 10x20. When I was browsing `snacks.nvim`, one of a handful of plugins that wrote earlier image logic for neovim, they defaulted to 9x18, hence why I picked that first. But I have no preference either way on what the default is.

Reply:
I'm not sure why anyone would choose 9x18 as a default (there may well be a good reason). I just know that 10x20 is the cell size of the original VT240 and VT330/340 terminals, so it's what real terminal emulators need to emulate if they want to be able to run software that targetted those devices.

Reply:
Is windows terminal hard coded to that? Do you maybe have a link to resources about how windows terminal does things? I am obviously also interested in that because of #28621, but don't have access to a windows machine at the moment to test things. I should really just spin it up in a VM though anyway.

Reply:
Yeah, at least for now Windows Terminal is hardcoded to 10x20, but it's possible there might come a time where they make that a configurable option. I personally don't think that's a good idea, and if it were up to me I wouldn't allow it, but it's not my decision to make.

So again I would recommend the use of size query sequences if at all possible. If you want to support third-party terminals like WezTerm, Contour, or Mintty, then that's essential. But falling back to 10x20 is better than nothing, because that should at least work on _some_ terminals.